src/Generics/SOP/Arbitrary.hs

```diff
@@ -25,13 +25,41 @@ import Generics.SOP
 --
 -- > instance Arbitrary T where
 -- >   arbitrary = garbitrary
+-- >   shrink    = gshrink
 --
 -- Note that currently no attempts are being made to generate arbitrary
 -- values of a particular size, and it is possible that this function
 -- diverges for recursive structures.
 --
 garbitrary :: forall a. (Generic a, All2 Arbitrary (Code a)) => Gen a
-garbitrary = liftM to $ hsequence =<< elements (apInjs_POP $ hcpure p arbitrary)
+garbitrary = fmap to $ oneof (map hsequence $ apInjs_POP $ hcpure p arbitrary)
   where
     p :: Proxy Arbitrary
     p = Proxy
+-- Note: we could generate all elements of sums and pick on with 'elements'.
+-- Instead we generate a NP Gen (Code a), and pick single generator with 'oneof'.
+
+-- | Generic shrink. See 'garbitrary'.
+--
+-- Note that this doesn't work very well for recursive structures:
+--
+-- >>> sort $ shrink ([1,2] :: [Int])
+-- [[],[0,2],[1],[1,0],[1,1],[2]]
+--
+-- >>> sort $ gshrink ([1,2] :: [Int])
+-- [[0,2],[1],[1,0],[1,1]]
+--
+gshrink :: forall a. (Generic a, All2 Arbitrary (Code a)) => a -> [a]
+gshrink
+    = map (to . SOP)
+    . hsequence'
+    . hcmap (Proxy :: Proxy (All Arbitrary)) gshrink'
+    . unSOP
+    . from
+  where
+    gshrink' :: forall ys. All Arbitrary ys => NP I ys -> ([] :.: NP I) ys
+    gshrink' Nil          = Comp $ []
+    gshrink' (I x :* Nil) = Comp $ fmap (\x' -> I x' :* Nil) $ shrink x
+    gshrink' (I x :* xs)  = Comp $
+        [ I x' :* xs  | x'  <- shrink x ] ++
+        [ I x  :* xs' | xs' <- unComp $ gshrink' xs ]
```

Comment:
The tuple instance hopefully clarifies:

```hs
instance (Arbitrary a, Arbitrary b)
      => Arbitrary (a,b)
 where
  arbitrary = liftM2 (,) arbitrary arbitrary

  shrink (x, y) =
       [ (x', y) | x' <- shrink x ]
    ++ [ (x, y') | y' <- shrink y 
```

E.g. we keep one component fixed and shrink another one. And then we do the same, reversed.

---

And now as I think properly, the singleton case isn't necessary